File: stripe/src/main/java/com/stripe/android/ApiRequest.java
package com.stripe.android;

import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.stripe.android.exception.InvalidRequestException;
import com.stripe.android.utils.ObjectUtils;

import org.json.JSONObject;

import java.io.UnsupportedEncodingException;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

/**
 * A class representing a Stripe API or Analytics request.
 */
final class ApiRequest extends StripeRequest {
    static final String MIME_TYPE = "application/x-www-form-urlencoded";

    static final String API_HOST = "https://api.stripe.com";

    @NonNull final Options options;
    @NonNull private final String mApiVersion;
    @Nullable private final AppInfo mAppInfo;

    ApiRequest(@NonNull Method method,
               @NonNull String url,
               @Nullable Map<String, ?> params,
               @NonNull Options options,
               @Nullable AppInfo appInfo) {
        super(method, url, params, MIME_TYPE);
        this.options = options;
        mApiVersion = ApiVersion.get().code;
        mAppInfo = appInfo;
    }

    @NonNull
    static ApiRequest createGet(@NonNull String url,
                                @NonNull Options options,
                                @Nullable AppInfo appInfo) {
        return new ApiRequest(Method.GET, url, null, options, appInfo);
    }

    @NonNull
    static ApiRequest createGet(@NonNull String url,
                                @NonNull Map<String, ?> params,
                                @NonNull Options options,
                                @Nullable AppInfo appInfo) {
        return new ApiRequest(Method.GET, url, params, options, appInfo);
    }

    @NonNull
    static ApiRequest createPost(@NonNull String url,
                                 @NonNull Options options,
                                 @Nullable AppInfo appInfo) {
        return new ApiRequest(Method.POST, url, null, options, appInfo);
    }

    @NonNull
    static ApiRequest createPost(@NonNull String url,
                                 @NonNull Map<String, ?> params,
                                 @NonNull Options options,
                                 @Nullable AppInfo appInfo) {
        return new ApiRequest(Method.POST, url, params, options, appInfo);
    }

    @NonNull
    static ApiRequest createDelete(@NonNull String url,
                                   @NonNull Options options,
                                   @Nullable AppInfo appInfo) {
        return new ApiRequest(Method.DELETE, url, null, options, appInfo);
    }

    @NonNull
    @Override
    Map<String, String> createHeaders() {
        final Map<String, String> headers = new HashMap<>();
        headers.put("Accept-Charset", CHARSET);
        headers.put("Accept", "application/json");
        headers.put("X-Stripe-Client-User-Agent", createStripeClientUserAgent());
        headers.put("Stripe-Version", mApiVersion);
        headers.put("Authorization",
                String.format(Locale.ENGLISH, "Bearer %s", options.apiKey));
        if (options.stripeAccount != null) {
            headers.put("Stripe-Account", options.stripeAccount);
        }
        return headers;
    }

    @NonNull
    private String createStripeClientUserAgent() {
        final Map<String, String> propertyMap = new HashMap<>();
        final String javaVersion = System.getProperty("java.version");
        if (javaVersion != null) {
            propertyMap.put("java.version", javaVersion);
        }
        propertyMap.put("os.name", "android");
        propertyMap.put("os.version", String.valueOf(Build.VERSION.SDK_INT));
        propertyMap.put("bindings.version", BuildConfig.VERSION_NAME);
        propertyMap.put("lang", "Java");
        propertyMap.put("publisher", "Stripe");
        if (mAppInfo != null) {
            propertyMap.putAll(mAppInfo.createClientHeaders());
        }

        return new JSONObject(propertyMap).toString();
    }

    @NonNull
    @Override
    String getUserAgent() {
        final StringBuilder userAgent = new StringBuilder(DEFAULT_USER_AGENT);
        if (mAppInfo != null) {
            userAgent
                    .append(" ")
                    .append(mAppInfo.toUserAgent());
        }
        return userAgent.toString();
    }

    @NonNull
    @Override
    byte[] getOutputBytes() throws UnsupportedEncodingException, InvalidRequestException {
        return createQuery().getBytes(CHARSET);
    }

    @Override
    public int hashCode() {
        return ObjectUtils.hash(getBaseHashCode(), options, mAppInfo);
    }

    @Override
    public boolean equals(@Nullable Object obj) {
        return super.equals(obj) ||
                (obj instanceof ApiRequest && typedEquals((ApiRequest) obj));
    }

    private boolean typedEquals(@NonNull ApiRequest obj) {
        return super.typedEquals(obj) &&
                ObjectUtils.equals(options, obj.options) &&
                ObjectUtils.equals(mAppInfo, obj.mAppInfo);
    }

    /**
     * Data class representing options for a Stripe API request.
     */
    static final class Options {
        @NonNull final String apiKey;
        @Nullable final String stripeAccount;

        @NonNull
        static Options create(@NonNull String apiKey) {
            return new Options(apiKey, null);
        }

        @NonNull
        static Options create(@NonNull String apiKey, @Nullable String stripeAccount) {
            return new Options(apiKey, stripeAccount);
        }

        private Options(
                @NonNull String apiKey,
                @Nullable String stripeAccount) {
            this.apiKey = new ApiKeyValidator().requireValid(apiKey);
            this.stripeAccount = stripeAccount;
        }

        @Override
        public int hashCode() {
            return ObjectUtils.hash(apiKey, stripeAccount);
        }

        @Override
        public boolean equals(@Nullable Object obj) {
            return super.equals(obj) || (obj instanceof Options && typedEquals((Options) obj));
        }

        private boolean typedEquals(@NonNull Options obj) {
            return ObjectUtils.equals(apiKey, obj.apiKey) &&
                    ObjectUtils.equals(stripeAccount, obj.stripeAccount);
        }
    }
}

File: stripe/src/main/java/com/stripe/android/ApiRequest.kt
package com.stripe.android

import android.os.Parcelable
import com.stripe.android.exception.InvalidRequestException
import java.io.UnsupportedEncodingException
import kotlinx.android.parcel.Parcelize

/**
 * A class representing a Stripe API or Analytics request.
 */
internal data class ApiRequest internal constructor(
    override val method: Method,
    override val baseUrl: String,
    override val params: Map<String, *>? = null,
    internal val options: Options,
    private val appInfo: AppInfo? = null,
    private val systemPropertySupplier: (String) -> String = DEFAULT_SYSTEM_PROPERTY_SUPPLIER,
    private val apiVersion: String = ApiVersion.get().code,
    private val sdkVersion: String = Stripe.VERSION
) : StripeRequest() {
    override val mimeType: MimeType = MimeType.Form

    override val headersFactory = RequestHeadersFactory.Api(
        options = options,
        appInfo = appInfo,
        systemPropertySupplier = systemPropertySupplier,
        apiVersion = apiVersion,
        sdkVersion = sdkVersion
    )

    override val body: String
        @Throws(UnsupportedEncodingException::class, InvalidRequestException::class)
        get() {
            return query
        }

    override fun toString(): String {
        return "${method.code} $baseUrl"
    }

    /**
     * Data class representing options for a Stripe API request.
     */
    @Parcelize
    internal data class Options internal constructor(
        val apiKey: String,
        internal val stripeAccount: String? = null,
        internal val idempotencyKey: String? = null
    ) : Parcelable {
        init {
            ApiKeyValidator().requireValid(apiKey)
        }
    }

    class Factory(
        private val appInfo: AppInfo? = null,
        private val apiVersion: String = ApiVersion.get().code,
        private val sdkVersion: String = Stripe.VERSION
    ) {
        fun createGet(
            url: String,
            options: Options,
            params: Map<String, *>? = null
        ): ApiRequest {
            return ApiRequest(
                Method.GET, url, params, options, appInfo,
                apiVersion = apiVersion,
                sdkVersion = sdkVersion
            )
        }

        fun createPost(
            url: String,
            options: Options,
            params: Map<String, *>? = null
        ): ApiRequest {
            return ApiRequest(
                Method.POST, url, params, options, appInfo,
                apiVersion = apiVersion,
                sdkVersion = sdkVersion
            )
        }

        fun createDelete(
            url: String,
            options: Options
        ): ApiRequest {
            return ApiRequest(
                Method.DELETE,
                url,
                options = options,
                appInfo = appInfo,
                apiVersion = apiVersion,
                sdkVersion = sdkVersion
            )
        }
    }

    internal companion object {
        internal const val API_HOST = "https://api.stripe.com"

        internal const val HEADER_STRIPE_CLIENT_USER_AGENT = "X-Stripe-Client-User-Agent"
    }
}
